cocotb/share/lib/embed/gpi_embed.cpp

```diff
@@ -227,6 +227,8 @@ extern "C" int embed_sim_init(int argc, char const * const * argv)
     if (pEventFn)
         return ret;
 
+    simulator_is_active = 1;
```

Comment:
Can you convince me this is the right place?

Reply:
The line place in the function, or the whole approach? If just the line placement in the function, it's kind of arbitrary, as long as it's set before we load into the Python entry point.

Reply:
The line in the function - ideally this would go right next to the code that initializes `registered_impls`

Reply:
That's not in the function? You mean this should be moved into the GPI, which is a different approach.

The GPI should only ever be loaded by the simulator, so we can assume once we are in the GPI, there is a valid simulator. Because we want to support loading the simulator module outside of the typical flow, the responsibility should rest with the simulator module to prevent itself from using the GPI when it knows it would be invalid.

Moving the flag into the GPI and adding a function to the GPI interface to return that flag means the user must check the function before using any interface. Robust design dictates we shouldn't trust the user and now we have to add error code handling to the entry points of the GPI.

In short, my approach is "No touchy!". I have tested using all the methods and types available on the simulator module from the terminal and ensured they do not segfault. If you think that warrants a test, I'd be happy to add one. I will also add a `simulator.is_running` function. No matter the approach that would be useful.

Reply:
> The GPI should only ever be loaded by the simulator, so we can assume once we are in the GPI, there is a valid simulator.

There are two different bits of the GPI - the C wrapper, which can be loaded by anyone, and the C++ code, which only gets instantiated when the simulator is entered.

I suppose what I'm advocating in the short term is a `gpi_has_registered_impl()` function implemented as `return registered_impls.size() == 0`. We already have `gpi_print_registered_impl()`, so it seems like a reasonable extension

Reply:
Actually, the `gpi_print_registered_impl` function already exposes the information you need via its return value, exposing that in the C extension API would do the trick.

Reply:
> the C wrapper, which can be loaded by anyone

I think that's the crux of the issue. My assumption is that most users know you can't load any library at any time, start using it, and expect it to not blow up; which allows us not do any of the work in the GPI. You seem to suggesting making the GPI interface tolerant against invalid usage in general, which is fine, it's just more work. #1844 seems related to that.

Reply:
I'm suggesting as a short term, we just add a method to the gpi that simply answers the question "will this blow up". We already have one at `gpi_print_registered_impl`, it just prints useless log messages we do not want.

With that method around, you could replace `if (simulator_is_active)` with `if(gpi_has_registered_impl())`. The function call cost (which ultimately just returns a data member of a global anyway) is insignificant compared to the python boundary cost.

Reply:
That's fine. I'll update this PR.

Reply:
Thanks!